corehq/apps/hqwebapp/models.py

```diff
@@ -53,3 +52,45 @@ def get_latest_alert(cls):
             return active_alerts[0]
         else:
             return ''
+
+
+class UserAgent(models.Model):
+    MAX_LENGTH = 255
+
+    value = models.CharField(max_length=MAX_LENGTH, db_index=True)
+
+
+class UserAccessLogManager(models.Manager):
+    def create(self, **obj_data):
+        user_agent = obj_data.pop('user_agent', '')
+        if user_agent:
+            user_agent = user_agent[:UserAgent.MAX_LENGTH]
+            agent_ref, _ = UserAgent.objects.get_or_create(value=user_agent)
+            obj_data['user_agent'] = agent_ref
+
+        return super().create(**obj_data)
+
+
+@architect.install('partition', type='range', subtype='date', constraint='month', column='timestamp')
+class UserAccessLog(models.Model):
+    TYPE_LOGIN = 'login'
+    TYPE_LOGOUT = 'logout'
+    TYPE_FAILURE = 'failure'
+
+    ACTIONS = (
+        (TYPE_LOGIN, 'Login'),
+        (TYPE_LOGOUT, 'Logout'),
+        (TYPE_FAILURE, 'Login Failure')
+    )
+
+    user_id = models.CharField(max_length=255, db_index=True)
+    action = models.CharField(max_length=20, choices=ACTIONS)
+    ip = models.GenericIPAddressField(blank=True, null=True)
+    user_agent = models.ForeignKey(UserAgent, null=True, on_delete=models.PROTECT)
+    path = models.CharField(max_length=255, blank=True)
+    timestamp = models.DateTimeField(default=datetime.utcnow)
```

Comment:
any reason not to use `auto_now` or `auto_now_add` instead of the `utcnow` default?

Reply:
Yeah -- I had `auto_now` originally, but didn't realize that means that the model itself is now solely responsible for timestamp generation (i.e. you can't override it). Given that this is security-related, I felt it was important to trace events back to exactly when they happened within the app. That necessitates passing in your own timestamp. The default was more a nicety that I don't expect to be used much.